instrumentation/opentelemetry-instrumentation-flask/tests/test_programmatic.py

```diff
@@ -243,3 +244,87 @@ def test_custom_span_name(self):
         span_list = self.memory_exporter.get_finished_spans()
         self.assertEqual(len(span_list), 1)
         self.assertEqual(span_list[0].name, "instrument-without-app")
+
+class TestProgrammaticHooks(
+    InstrumentationTest, TestBase, WsgiTestBase
+):
+    def setUp(self):
+        super().setUp()
+
+        hook_headers = (
+            "hook_attr",
+            "hello otel",
+        )
+
+        def request_hook(span, environ):
+            span.update_name("name from hook")
+
+        def response_hook(span, environ, response_headers):
+            span.set_attribute("hook_attr", "hello world")
+            response_headers.append(hook_headers)
+        
+        self.app = Flask(__name__)
+
+        FlaskInstrumentor().instrument_app(
+            self.app, otel_request_hook=request_hook, otel_response_hook=response_hook
+        )
+
+        self._common_initialization()
+    
+    def tearDown(self):
+        super().tearDown()
+        with self.disable_logging():
+            FlaskInstrumentor().uninstrument_app(self.app)
+    
+    def test_hooks(self):
+        expected_attrs = expected_attributes(
+            {"http.target": "/hello/123", "http.route": "/hello/<int:helloid>", "hook_attr":"hello world"}
+        )
+
+        self.client.get("/hello/123")
+        span_list = self.memory_exporter.get_finished_spans()
+        self.assertEqual(len(span_list), 1)
+        self.assertEqual(span_list[0].name, "name from hook")
+        self.assertEqual(span_list[0].attributes, expected_attrs)
+
+class TestProgrammaticHooksWithoutApp(
+    InstrumentationTest, TestBase, WsgiTestBase
+):
+    def setUp(self):
+        super().setUp()
+
+        hook_headers = (
+            "hook_attr",
+            "hello otel without app",
+        )
+
+        def request_hook(span, environ):
+            span.update_name("without app")
+
+        def response_hook(span, environ, response_headers):
+            span.set_attribute("hook_attr", "hello world without app")
+            response_headers.append(hook_headers)
+
+        FlaskInstrumentor().instrument(otel_request_hook=request_hook, otel_response_hook=response_hook)
+        # pylint: disable=import-outside-toplevel,reimported,redefined-outer-name
+        from flask import Flask
```

Comment:
isn't flask being imported globally above? why do we need this?

Reply:
Hey Owais, so if I don't import Flask here, no span will be created. I'm not sure exactly why this is the case but when I removed that import line, the span was no longer being created.

Reply:
I think you'll be able to solve it by creating your flask app instance after calling instrument(). So for each test case, first call `.instument()` and then create flask app instance.

Reply:
Hey Owais, if I'm not mistaken, aren't I already calling .instrument() before I import Flask and create the app instance?

Reply:
I don't see the app being created anywhere in this method. May be it is being created in the super-class in which case it would be created before since super().setUp() is being called before instrument.

Reply:
Hey Owais, for this test, we're testing the hooks without the App being created. I'm not entirely sure how this is supposed to work, but previously there was a test for ` TestProgrammaticCustomSpanNameWithoutApp`, which did the same. I believe that @lonewolf3739 may have a better understanding of what this is used for.

Reply:
@NickSulistio Sorry for getting back late. I think If you do import `from flask import Flask` first and next call `FlaskInstrumentor().instrument()`then instrumentation doesn't happen because the way patching is done; I remember this was the reason but I may be wrong.

Reply:
If this is a limitation with the instrumentation itself, I think we can live with it for now and try to improve the instrumentation later.